### PR TITLE
Add CUDA implementations

### PR DESCRIPTION
**Pull request by kakwok**

### CODE REVIEW

**Comment by alexstrel** [comment]
@kakwok Do you think to use python scripting for the build process in the future? (Not sure that this will be convenient if the application will be extended). 
Also, current script seems not allowing for the gpu architecture selection.

**Comment by kakwok** [comment]
@alexstrel I think the python script should be sufficient to handle the different architectures that we have in mind for now. 

Good point about the GPU architecture selection, we can add that to the script.  Since it's quite static and specific to the test environment, you can change the compile command by hand if you need the feature now.

**Comment by makortel** [comment]
> (Not sure that this will be convenient if the application will be extended).

@alexstrel Could you elaborate what kind of "extension of application" are you thinking about?

The main motivation for the script comes from all knobs being compile-time macros (and from the code it would be simpler to keep them that way), and therefore any parameter scan ties the compilation and running together.

**Comment by alexstrel** [comment]
@kakwok regarding the performance critical parameters, i.e.   ` --num_streams`, ` --threadsperblock`, `  --threadsperblockx`, `  --threadsperblocky` and `  --blockspergrid` ( with `--bsize` parameter relevant to SIMD-capable architectures). If understood you correctly they are selected/redefined via the script as user's external arguments. 
That is the user is responsible for tuning the application, is it correct? Instead, one may include autotuning mechanism directly into the application, say the first time the user run the application in 2 stages :  within the first stage will  run the tuner, cashes the optimal params to a specified file and , within the second stage, just use the pre-defined params for actual benchmarking.  
This goes, of course, beyond the scope of the current PR.  
Regarding the remaining arguments, I wonder why this should be passed via the script? Is it necessary for them from the performance point of view to be defined at compile time?

@makortel I meant that the package will be similar to p2z with possible option to use a variety of external libraries (e.g. eigen, oneDPL).

File: src/propagate-tor-test_cuda.cu
Comment (makortel) [inline src/propagate-tor-test_cuda.cu]: These transfers are unnecessary, right? (since the first copy transfers them all)

File: src/propagate-tor-test_cuda.cu
Comment (makortel) [inline src/propagate-tor-test_cuda.cu]: Same here

File: src/propagate-tor-test_cuda.cu
Comment (makortel) [inline src/propagate-tor-test_cuda.cu]: And here

File: src/propagate-tor-test_cuda.cu
Comment (makortel) [inline src/propagate-tor-test_cuda.cu:932]: This makes me want a clang-format.

File: src/propagate-tor-test_cuda.cu
Comment (makortel) [inline src/propagate-tor-test_cuda.cu:935]: This would actually need a synchronization point to start to clock only after the transfers are complete
```suggestion
   cudaDeviceSynchronize();
   auto wall_start = std::chrono::high_resolution_clock::now();
```
(and it should really be done better but I'd leave that to a bit later time)

File: src/propagate-tor-test_cuda_v2.cu
Comment (makortel) [inline src/propagate-tor-test_cuda_v2.cu:924]:
```suggestion
   cudaDeviceSynchronize();
   auto wall_start = std::chrono::high_resolution_clock::now();
```

File: src/propagate-tor-test_cuda_v3.cu
Comment (makortel) [inline src/propagate-tor-test_cuda_v3.cu:943]:
```suggestion
   cudaDeviceSynchronize();
   auto wall_start = std::chrono::high_resolution_clock::now();
```

**Review by makortel — CHANGES_REQUESTED** [review]
I think tidying up the transfers is not strictly necessary for this version (since we time only the kernel).

It would be clearer if the default parameters on top of each file would correspond the best performing point found so far (including setting the number of streams to 1), especially thinking those who would prefer to compile files without the scanning script.

It would also be good if the scope of the time/throughput measurement could be controlled with macro(s), e.g. cases "kernel only", "kernel+input transfer", "kernel+output transfer", "kernel + all transfers". I'd also leave it later (and could volunteer to implement during the hackathon). 

In summary, I'd ask before merging
- add the missing `cudaDeviceSynchronize()` calls (implies repeating the throughput scans)
- set the default parameter values to the current best

**Comment by makortel** [comment]
@alexstrel 

> I meant that the package will be similar to p2z with possible option to use a variety of external libraries (e.g. eigen, oneDPL).

Ok. For libraries like Kokkos or Alpaka I'd make the script download and compile (if necessary) the library, similar to what [cms-patatrack/pixeltrack-standalone](https://github.com/cms-patatrack/pixeltrack-standalone/) does.

For others, I don't see what value some other build system (e.g. Make, CMake) would bring because we're building independent source files. Ok, for e.g. Kokkos CMake would bring some value by avoiding the need to express the build commands explicitly in the script. But even then it would be useful to have some script to drive scanning of throughput across the various parameters, and as long as the parameters are set as compile time macros, a script for both compiling and running the program(s) seems the most convenient option to me.

**Comment by kakwok** [comment]
@makortel 

Synchronization before the start of the kernel and redundant transferAsync commands are fixed in this commit:  b05cc28
The update gives a constant offset of ~20% higher throughput for CUDA v1 as seen below:
image

Default values for parameters are also updated in 6487e57

**Review by makortel — APPROVED** [review]

**Pull request merged**
